src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs

```diff
@@ -463,16 +467,28 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
             // |  String Data (Length octets)  |
             // +-------------------------------+
 
-            if (destination.Length != 0)
+            if (destination.Length > value.Length)
             {
                 destination[0] = 0; // TODO: Use Huffman encoding
-                if (IntegerEncoder.Encode(value.Length, 7, destination, out int integerLength))
+
+                int encodedStringLength = valueEncoding is null || valueEncoding.IsSingleByte ? value.Length : valueEncoding.GetByteCount(value);
```

Comment:
`IsSingleByte` determines if the encoding itself has a multi-byte encoding. It does not indicate that there is a 1:1 mapping from UTF-16 code units to encoded bytes.

```suggestion
                int encodedStringLength = valueEncoding is null ? value.Length : valueEncoding.GetByteCount(value);
```

Reply:
Changed to `valueEncoding is null || ReferenceEquals(valueEncoding, Encoding.Latin1)`